### PR TITLE
HDDS-10446. Refactoring Node2ObjectsMap, Node2PipelineMap, Node2ContainerMap

### PR DESCRIPTION
**Pull request by Montura**
There is a base class [`org.apache.hadoop.hdds.scm.node.states.Node2ObjectsMap`](https://github.com/apache/ozone/blob/master/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/states/Node2ObjectsMap.java#L40) for [`Node2PipelineMap`](https://github.com/apache/ozone/blob/master/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/states/Node2PipelineMap.java#L37) and [`Node2ContainerMap`](https://github.com/apache/ozone/blob/master/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/states/Node2ContainerMap.java#L37).

It is redundant because of:
1. `Node2ContainerMap` is used only in tests
2. `Node2PipelineMap` doesn't  need inheritance at all

Moreover `Node2PipelineMap` contains `ConcurrentHashMap<UUID, Set<PipelineID>>` and two `synchronized` methods:
* `void addPipeline(Pipeline pipeline)` -> [compute if absent for concurrent hash map](https://github.com/apache/ozone/blob/master/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/states/Node2PipelineMap.java#L68)
* `void removePipeline(Pipeline pipeline)` -> [compute if present for concurrent hash map](https://github.com/apache/ozone/blob/master/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/states/Node2PipelineMap.java#L76)

I suppose `synchronized` is redundant here for these two methods, see [comment](https://github.com/apache/ozone/pull/6303#issuecomment-1977415944)  

## What changes were proposed in this pull request?
1. Remove `org.apache.hadoop.hdds.scm.node.states.Node2ObjectsMap` class
2. Move `Node2ContainerMap` to tests
4. Don't use `synchronized` for `Node2PipelineMap::addPipeline` and `Node2PipelineMap::removePipeline`

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-10446

## How was this patch tested?

Use standalone tests

### CODE REVIEW

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/node/states/TestNode2ContainerMap.java [hunk 1]
```diff
@@ -20,6 +20,7 @@
 package org.apache.hadoop.hdds.scm.node.states;
 
 import org.apache.hadoop.hdds.scm.container.ContainerID;
+import org.apache.hadoop.hdds.scm.container.Node2ContainerMap;
```
Comment (adoroszlai) [inline hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/node/states/TestNode2ContainerMap.java:23]: Please also move the test to `package org.apache.hadoop.hdds.scm.container`.

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/Node2ContainerMap.java [hunk 2]
```diff
@@ -16,37 +16,47 @@
...
 
 import static org.apache.hadoop.hdds.scm.exceptions.SCMException.ResultCodes.DUPLICATE_DATANODE;
+import static org.apache.hadoop.hdds.scm.exceptions.SCMException.ResultCodes.NO_SUCH_DATANODE;
 
 /**
  * This data structure maintains the list of containers that is on a datanode.
  * This information is built from the DN container reports.
  */
-public class Node2ObjectsMap<T> {
+public class Node2ContainerMap {
```
Comment (adoroszlai) [inline hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/Node2ContainerMap.java:40]: After moving the test, this class can be package-private.

```suggestion
class Node2ContainerMap {
```

**Review by adoroszlai — COMMENTED** [review]
Thanks @Montura for the patch.

**Comment by szetszwo** [comment]
> Moreover Node2PipelineMap contains ConcurrentHashMap<UUID, Set<PipelineID>> and two synchronized methods ...

Since there are multiple nodes, removing `synchronized` can lead to an inconsistent state.  Suppose Pipeline `p` has datanodes [A, B, C] and we have the following two calls:
1. addPipeline(p)
2. removePipeline(p)

With `synchronized`, all datanodes A, B, C will be added and then removed.

Without `synchronized`, it may
- add A, B
- remove A, B, C
- add C

Then, C is not removed.

**Comment by szetszwo** [comment]
> ...  methods  SCMNodeManager::addPipeline and SCMNodeManager::removePipeline are called under write_lock from PipelineStateManagerImpl,  ...

I see.  The methods are protected by an external locking in `PipelineStateManagerImpl` but not the `computeIfAbsent`/`computeIfPresent` methods.  Then, removing `synchronized` is okay.  Thanks for checking!

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.  It is a good cleanup.

**Comment by adoroszlai** [comment]
Thanks @Montura for the patch, @szetszwo for the review.

**Pull request merged**